src/tests/async/collectible-alc.cs

```diff
@@ -17,14 +17,20 @@ public static void TestEntryPoint()
         AsyncEntryPoint().Wait();
     }
 
+    [MethodImpl(MethodImplOptions.NoInlining)]
+    private static void CollectOnce()
```

Comment:
This is a workaround for an assert failure. 
The assert is caused by this PR (as in - it does not happen without the changes, but I am not familiar enough with JIT codebase to tell right away how it is related to the changes. Seems like something related to PInvoke setup. . . 

without the workaround: 
```
Assert failure(PID 55592 [0x0000d928], Thread: 16608 [0x40e0]): Assertion failed 'comp->fgFirstBBisScratch()' in 'Async2CollectibleAlc:AsyncEntryPoint()' during 'Lowering nodeinfo' (IL size 49; hash 0x76f7a2c6; FullOpts)

    File: E:\dotnetA2\runtimelab\src\coreclr\jit\lower.cpp:5905
    Image: E:\dotnetA2\runtimelab\artifacts\tests\coreclr\windows.x64.Checked\Tests\Core_Root\corerun.exe

```

Reply:
This is not a new issue, it reproduces also before this PR with `DOTNET_TieredCompilation=0`, so we can ignore this particular issue.

However, the fact that it reproduces here means that tiered compilation is not working with the new scheme. The new thunks seem to be handled differently from metadata functions with respect to PGO, optimizations, etc.. Can you take a look at that?

Reply:
Fix for the pinvoke issue is up in #2847.

Reply:
> The new thunks seem to be handled differently from metadata functions with respect to PGO, optimizations, etc.. Can you take a look at that?

The thunks were explicitly excluded from tiering. perhaps because thunk methods are too small. 
Exclusion is one of a few places that should not switch to `IsAsyncHelperMethod()` and should continue using `IsAsyncThunkMethod`.

Reply:
By default everything that was using `IsAsyncThunkMethod()` switched to `IsAsyncHelperMethod()`, but in this case it was incorrect and effectively disabled tiering of runtime async methods.

Reply:
I was trying to validate that PGO works now, but the new Roslyn prototype fails to compile my test example:
```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.

using System;
using System.Diagnostics;
using System.Linq;
using System.Threading.Tasks;
using Xunit;

public class Async2Pgo
{
    [Fact]
    public static void EntryPoint()
    {
        AsyncEntryPoint().Wait();
    }
    
    internal static async2 Task<int> AsyncEntryPoint()
    {
        int[] arr = Enumerable.Range(0, 100_000).ToArray();

        int sum = 0;
        for (int i = 0; i < 4; i++)
        {
            for (int j = 0; j < 100; j++)
                sum += await AggregateDelegateAsync(arr, new AggregateSum(), 0);
                
            await Task.Delay(100);
        }
        
        return sum;
    }
    
    private class AggregateSum : I<int>
    {
#pragma warning disable CS1998
        public async2 Task<int> Aggregate(int a, int b) => a + b;
    }
    
    public interface I<T>
    {
        public Task<T> Aggregate(T seed, T val);
    }
    
    public static async2 Task<T> AggregateDelegateAsync<T>(T[] arr, I<T> aggregate, T seed)
    {
        foreach (T val in arr)
            seed = await aggregate.Aggregate(seed, val);
            
        return seed;
    }
}
```

```
C:\Users\Jakob\.nuget\packages\microsoft.net.compilers.toolset\4.13.0-async-11\tasks\netcore\Microsoft.CSharp.Core.targets(89,5): er
ror : Process terminated. System.NullReferenceException: Object reference not set to an instance of an object. [C:\dev\dotnet\runtim
elab\src\tests\async\pgo.csproj]
C:\Users\Jakob\.nuget\packages\microsoft.net.compilers.toolset\4.13.0-async-11\tasks\netcore\Microsoft.CSharp.Core.targets(89,5): er
ror :    at Microsoft.CodeAnalysis.CSharp.Symbols.Async2ThunkForAsyncMethod.get_ReturnTypeWithAnnotations() in C:\dev\dotnet\async-r
oslyn-repo\src\Compilers\CSharp\Portable\Symbols\AsyncThunks.cs:line 83 [C:\dev\dotnet\runtimelab\src\tests\async\pgo.csproj]
C:\Users\Jakob\.nuget\packages\microsoft.net.compilers.toolset\4.13.0-async-11\tasks\netcore\Microsoft.CSharp.Core.targets(89,5): er
ror :    at Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol.get_ReturnType() in C:\dev\dotnet\async-roslyn-repo\src\Compilers\CSh
arp\Portable\Symbols\MethodSymbol.cs:line 251 [C:\dev\dotnet\runtimelab\src\tests\async\pgo.csproj]
C:\Users\Jakob\.nuget\packages\microsoft.net.compilers.toolset\4.13.0-async-11\tasks\netcore\Microsoft.CSharp.Core.targets(89,5): er
ror :    at Microsoft.CodeAnalysis.CSharp.Symbols.MethodSymbol.Microsoft.Cci.ISignature.GetType(EmitContext context) in C:\dev\dotne
t\async-roslyn-repo\src\Compilers\CSharp\Portable\Emitter\Model\MethodSymbolAdapter.cs:line 245 [C:\dev\dotnet\runtimelab\src\tests\
async\pgo.csproj]
...
```